### DISCUSSION TITLE
Providing more information when throwing an error in the case of `undefined query data being returned`

### DISCUSSION THREAD

**Discussion by samueljoli** [TanStack/query#6948] [Q&A]
Currently, in `v4` when no data is being returned from a query it passes an error like so to an `onError` callback:

```ts
        if (typeof data === 'undefined') {
          if (process.env.NODE_ENV !== 'production') {
            this.logger.error(
              `Query data cannot be undefined. Please make sure to return a value other than undefined from your query function. Affected query key: ${this.queryHash}`,
            )
          }
          onError(new Error('undefined') as any)
          return
        }
```

I don't typically have my logger turned on during testing so it took me a second to figure out what the issue was. It turned out to be an improperly configured url being passed to `jest-when` hence the api call wasn't being properly mocked to return the expected results. I guess I should have just turned on my logger and this would have surfaced. I would imagine this being the recommended debugging protocol, but I wonder if it would make sense to provide just a tad bit more context in the error being thrown for those of use who don't instinctively turn on our loggers during testing.

**suggestion:**
```ts
onError(new Error('Query data cannot be undefined') as any)
```

Related to: https://github.com/TanStack/query/pull/4218

**Comment by TkDodo** [comment]
one problem is that those rather long strings stay in the final bundle as-is, so we've tried to keep it to a minimum and rather log during dev mode. Now we could throw a different error with more info in dev mode if that helps:

```
        if (typeof data === 'undefined') {
          if (process.env.NODE_ENV !== 'production') {
            const message = `Query data cannot be undefined. Please make sure to return a value other than undefined from your query function. Affected query key: ${this.queryHash}`
            this.logger.error(message)
            onError(new Error(message) as any)
            return
          }
          onError(new Error('undefined') as any)
          return
        }
```

because this will be removed by bundlers. However, then we have different errors in dev and prod, so not sure if that doesn't have any drawbacks 😅

**Reply by samueljoli** [reply]
😅 Ahh I'm not aware of any drawbacks of differing messages based on env

**Reply by TkDodo** [reply]
Okay then please PR this change
